radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/proxy/ProxyCertificateManager.java

```diff
@@ -0,0 +1,309 @@
+			}
+		}
+	}
+
+	private void verifyAndUpdateProxyCertificates(NodeId peer, ImmutableSet<ProxyCertificate> proxyCertificates) {
+		final var verifiedCertificates = proxyCertificates.stream()
+			.map(cert -> this.verifyProxyCertificate(peer, cert))
+			.filter(Optional::isPresent)
+			.map(Optional::get)
+			.collect(ImmutableSet.toImmutableSet());
+
+		if (verifiedCertificates.size() == proxyCertificates.size()) {
+			verifiedCertificates.forEach(verifiedCertificate -> {
+				final var verifiedChannelsForSigner = this.verifiedProxies.computeIfAbsent(
+					verifiedCertificate.signer(),
+					unused -> Sets.newConcurrentHashSet()
+				);
+				verifiedChannelsForSigner.add(verifiedCertificate);
+			});
+			this.addressBook.updateProxyCertificates(peer, verifiedCertificates);
+		} else {
+			final var banDuration = Duration.ofMinutes(15);
+			log.warn(
+				"Received invalid proxy certificate from {}, banning for {}",
+				addressing.forNodes().of(peer.getPublicKey()),
+				banDuration
+			);
+			peerControl.banPeer(peer, banDuration, "Received invalid proxy certificate");
+		}
+	}
+
+	public EventProcessor<RenewIssuedProxyCertificatesTrigger> renewIssuedProxyCertificatesTriggerProcessor() {
+		return unused -> {
+			// renew if cert expired or is about to expire in less than a minute
+			final var minExpirationTime = Instant.now().plus(Duration.ofMinutes(1));
+			final var certsToRenew = this.issuedProxyCertificates.entrySet().stream()
+				.filter(e -> minExpirationTime.isAfter(Instant.ofEpochMilli(e.getValue().getData().expiresAt())))
+				.map(Map.Entry::getKey)
+				.collect(ImmutableSet.toImmutableSet());
+
+			certsToRenew.forEach(this::issueProxyCertificate);
+		};
+	}
+
+	private Optional<VerifiedProxyCertificate> verifyProxyCertificate(
+		NodeId expectedGrantee,
+		ProxyCertificate proxyCertificate
+	) {
+		return proxyCertificate.verify()
+			.filter(cert -> cert.grantee().equals(expectedGrantee))
+			.filter(cert -> cert.networkId() == networkId);
+	}
+
+	private void issueProxyCertificate(NodeId nodeId) {
+		log.debug("Issuing a proxy certificate for {}", addressing.forNodes().of(nodeId.getPublicKey()));
+		final var cert = issueProxyCertificate(nodeId, config.issuedProxyCertificateValidityDuration());
+		this.issuedProxyCertificates.put(nodeId, cert);
+		grantedProxyCertificateDispatcher.dispatch(
+			BFTNode.create(nodeId.getPublicKey()),
+			GrantedProxyCertificate.create(cert)
+		);
+	}
+
+	private ProxyCertificate issueProxyCertificate(NodeId nodeId, Duration validityDuration) {
+		final var expiresAt = Instant.now().plus(validityDuration).toEpochMilli();
+		final var data = ProxyCertificateData.create(nodeId, expiresAt, networkId);
+		final var signature = ecKeyOps.sign(data.hashToSign().asBytes());
+		return ProxyCertificate.create(data, signature);
+	}
+
+	public RemoteEventProcessor<GrantedProxyCertificate> grantedProxyCertificateEventProcessor() {
+		return (peer, ev) -> {
+			synchronized (lock) {
```

Comment:
Code is somewhat too long and complex for the lambda. I think it's worth to split lambda into smaller methods.